src/Extensions/PxPayExtension.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace SilverStripe\Omnipay\Extensions;
+
+use SilverStripe\Core\Extension;
+
+/**
+ * Fix request for PxPay.
+ *
+ * To apply this extension use the following configuration.
+ *
+ * SilverStripe\Omnipay\Service\PaymentService:
+ *  extensions:
+ *    - SilverStripe\Omnipay\Extensions\PxPayExtension
+ *
+ * SilverStripe\Omnipay\PaymentGatewayController:
+ *  extensions:
+ *    - SilverStripe\Omnipay\Extensions\PxPayExtension
+ */
+class PxPayExtension extends Extension
+{
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforePurchase(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeAuthorize(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeCompletePurchase(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param array $gatewayData
+     */
+    public function onBeforeCompleteAuthorize(array &$gatewayData)
+    {
+        $this->fixTransactionId($gatewayData);
+    }
+
+    /**
+     * @param HTTPRequest $request The current request
+     * @param string $gateway the gateway name.
+     *
+     * @return Iterable
+     */
+    public function updatePaymentFromRequest($request, $gateway = null)
+    {
+        if ($gateway == 'PaymentExpress_PxPay') {
+            $identifier = $request->param('Identifier');
+
+            if ($identifier && strlen($identifier) == 16) {
+                // identifier in PxPay is limited to 16 characters so can match
+                // on the start as long as we have 16 characters (i.e matching
+                // on `1` would not be secure).
+                //
+                // Identifiers are normally 30 characters.
+                return Payment::get()->filter([
```

Comment:
This extension hook should return a `Payment` or `null`, therefore it's better to use `find` instead of `filter` here. Or append `->first()` after the filter.

Reply:
Seems like `find` only works with a single key/value pair… so `filter(…)->first()` is the only option.

Reply:
Sorry I was misled by the fact that the callee loops overs the response but I figure that's how `extend` returns the results.

Reply:
Yeah exactly. Otherwise it would be an array in an array